src/optimizer/memo.cpp

```diff
@@ -43,8 +43,6 @@ GroupExpression *Memo::InsertExpression(std::shared_ptr<GroupExpression> gexpr,
   auto it = group_expressions_.find(gexpr.get());
 
   if (it != group_expressions_.end()) {
-    PELOTON_ASSERT(target_group == UNDEFINED_GROUP ||
-           target_group == (*it)->GetGroupID());
```

Comment:
I debugged for a while and reached the conclusion that this assert was incorrect. From my understanding, it doesn't make sense that if we create a group expression, we search for it in the memo table, and we find an existing equal expression, that we should already know it's `GroupID` when its passed in as `target_group`. Consider two possible join trees (AxB)x(CxD) and Ax(Bx(CxD)). In these two cases, both of the cases will generate a group expression for (CxD) independently. The first will generate it from the group Bx(CxD), but the second will generate it from the group (CxD).

Reply:
Yes after thinking I think you're correct. A case would be from some transformation we get `Ax(BxC)` and insert it into the memo, then from another transformation we get `(AxB)xC` and decide it should go into another group but actually they should be in the same group. Notice that we do not generate all logical expressions at once now so this is possible. Later we'll get this assert failure after applying`Ax(BxC) -> (AxB)xC`. So to solve this we need to merge groups in the memo.

Reply:
I actually think the assertion is fine... I think the fundamental problem is what you have discussed in Slack: the logical and physical join for the same children should be evaluated to be equal. But maybe i’m missing something. Let’s discuss today.